### PR TITLE
Branches - Eve

### PR DESCRIPTION
**Pull request by tofuandeve**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | If imagining Rails is a restaurant, then Model is the kitchen, including chefs, kitchen staff, structure of the kitchen and ingredients. Model deals with the logic of processing an order, finding the right ingredients and tools and making the right dish from an order.
Describe in your own words what the Controller is doing in Rails | The servers and expeditors, Controller directs traffic in the restaurant: routing orders to the right actions, and plating the dishes, and then routing ready dishes back to customers
Describe in your own words what the View is doing in Rails | View is the presentation of a dish to customers
Describe an edge-case controller test you wrote | I tested when .save fails due to invalid input of data (missing data, empty data, data is empty string) when creating/updating task.
What is the purpose of using strong params? (i.e. the params method in the controller) | To ensure that users cannot update sensitive model attributes that they should not be updating
How are Rails migrations related to Rails models? | Migration is the way Rails provides us to maintain and update models
Describe one area of Rails that are still unclear on | What will go inside `class Task < ApplicationRecord` in task.rb? Why do we talk a lot about model but the class Task in models folder has not been discussed?

### CODE REVIEW

**Comment by tildeee** [comment]
# Task List
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Appropriate Git Usage with no extraneous files checked in | x
Answered comprehension questions | x
Successfully handles: Index, Show | x
Index & Show tests pass | x
Successfully handles: New, Create | x
New & Create tests pass | x
Successfully handles: Edit, Update | x
Tests for Edit & Update test valid & invalid task ids | x
Successfully handles: Destroy, Task Complete | x
Tests for Destroy & Task Complete include tests for valid and invalid task ids | x
Routes follow RESTful conventions | x
Uses named routes (like `_path`) | x
**Overall** |



Great work on this project, Eve! Your code looks great-- the controller code looks great and the tests are thorough and detailed.


I don't have too many comments on this project; just a few questions for me to learn and some tiny suggestions


That being said, great work on this project. Keep it up!

File: app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,87 @@
+class TasksController < ApplicationController
+  def index
+    @tasks = Task.all
+  end
+  
+  def show
+    task_id_raw = params[:id]
+    task_id = task_id_raw.to_i
+    @task = Task.find_by(id: task_id)
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:9]: Just a nice-to-know: Actually, now that we have ActiveRecord up and running, ActiveRecord is very forgiving and we don't need to convert to integer... Aka, changing this code to this line still works:

```ruby
@task = Task.find_by(id: task_id_raw)
```

File: app/controllers/tasks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        redirect_to new_task_path
+        return
+      end 
+      
+      redirect_to task_path(@task.id)   
+      return  
+    end
+    
+    redirect_to new_task_path
+    return
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:35]: Nice! Your code doesn't quite follow the pattern at Ada that we teach (aka you prefer using `begin ... rescue` and exception handling to handle edge cases), but this code looks fantastic and works super well. I have no complaints-- just be mindful when you're on teams to see what style your team wants to go with.

File: test/controllers/tasks_controller_test.rb [hunk 3]
```diff
@@ -71,56 +64,240 @@
...
+          name: "updated task",
+          description: "updated task description", 
+          completion_date: nil
+        }
+      }
     end
 
+    it "can update an existing task" do
+      existing_task = Task.find_by(id: task[:id])
+      expect _(Task.count).must_equal 1
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:150]: this is super interesting syntax; I actually don't know what this is! I'll probably ask you at some point, since I haven't seen this before.

Comment (tofuandeve) [reply test/controllers/tasks_controller_test.rb:150]: @tildeee  Oh this is one of the weird things that minitest suggested me to do. Instead of ``` expect (obj).must_equal value``` minitest complains that this should be replaced with ``` expect _(obj).must_equal value``` so that the test won't fail on minitest 6. I had too many warnings about this I couldn't see where the stack trace was for my errors so I had to change my tests to use this syntax.
